Implement Brigadier Mojang API

aikar opened the pull request:
This is the start of a new module for Paper to add support for API's
that interface Mojang API's directly.

This allows us to version properly by MC version incase Mojang makes any major breaking changes.

It also lets us separate Mojang API's from Paper-API so our downstream friends at Glowstone
will not have to worry about Mojang code.

Adds AsyncPlayerSendCommandsEvent
  - Allows modifying on a per command basis what command data they see.

Adds CommandRegisteredEvent
  - Allows manipulating the CommandNode to add more children/metadata for the client

JavierKasunic commented on Paper-MojangAPI/src/main/java/com/destroystokyo/paper/event/brigadier/AsyncPlayerSendCommands.java:
```diff
@@ -0,0 +1,54 @@
+ *
+ * If your logic is not safe to run asynchronously, only react to the synchronous version.
+ */
+public class AsyncPlayerSendCommands extends PlayerEvent {
```
Wasn't this supposed to be called AsyncPlayerSendCommands**Event**?

aikar replied:
good catch

MiniDigger commented on Paper-MojangAPI/src/main/java/com/destroystokyo/paper/event/brigadier/AsyncPlayerSendCommandsEvent.java:
```diff
@@ -0,0 +1,64 @@
+ * If for some reason we are unable to send this asynchronously in the future, only the sync method will fire.
+ *
+ * Your logic should look like this:
+ * if (event.isAsync() || (!event.isAsync() && !event.hasFiredAsync())) { do stuff }
```
I know this is pseudo code, but for easy copy and pasting, this and the reference in line 20 should prolly say `isAsynchronous` since thats the proper name of the method.

IntelliJ also reminded me, that this can be simplified to `event.isAsynchronous() || !event.hasFiredAsync()`

MiniDigger reviewed (commented):
any reason paper-mojangapi wasnt added to papers parent pom?

MiniDigger reviewed (approved):
Heavily tested this with ACF, its working as expected and exposes everything that should be needed.
This makes working with brigadier so much easier, as bukkit no longer gets into your way and overrides stuff.
Now we can write plugins that easy uses brigadier (for example, via ACF) and ppl have yet another reason to push paper :D

chickeneer commented on Spigot-API-Patches/0001-POM-changes.patch:
```diff
@@ -110,6 +110,19 @@ index 8964b9e33..ce09baeab 100644
+     }
+ 
+-    PlayerEvent(@NotNull final Player who, boolean async) {
++    public PlayerEvent(@NotNull final Player who, boolean async) { // Paper - wtf?
```
I do not think this change is in the correct patch?

aikar replied:
I didn't want to make a new patch for it lol

aikar replied:
though i was thinking we already had other crap in pom patch, but thats on -Server

aikar committed:
Implement Brigadier Mojang API

This is the start of a new module for Paper to add support for API's
that interface Mojang API's directly.

This allows us to version properly by MC version incase Mojang makes any major breaking changes.

It also lets us separate Mojang API's from Paper-API so our downstream friends at Glowstone
will not have to worry about Mojang code.

Adds AsyncPlayerSendCommandsEvent
  - Allows modifying on a per command basis what command data they see.

Adds CommandRegisteredEvent
  - Allows manipulating the CommandNode to add more children/metadata for the client
